src/stream/xml_writer.rs

```diff
@@ -157,28 +145,65 @@ impl<W: Write> XmlWriter<W> {
             Ok(())
         })
     }
+
+    fn handle_pending_collection(&mut self) -> Result<(), Error> {
+        if let Some(PendingCollection::Array) = self.pending_collection {
+            self.pending_collection = None;
+            let res = self.write_value_event(EventKind::StartArray, |this| {
+                this.start_element("array")?;
+                this.stack.push(Element::Array);
+                Ok(())
+            });
+            res
+        } else if let Some(PendingCollection::Dictionary) = self.pending_collection {
+            self.pending_collection = None;
+            let res = self.write_value_event(EventKind::StartDictionary, |this| {
+                this.start_element("dict")?;
+                this.stack.push(Element::Dictionary);
+                this.expecting_key = true;
+                Ok(())
+            });
+            res
+        } else {
+            Ok(())
+        }
+    }
 }
 
 impl<W: Write> Writer for XmlWriter<W> {
     fn write_start_array(&mut self, _len: Option<u64>) -> Result<(), Error> {
-        self.write_value_event(EventKind::StartArray, |this| {
-            this.start_element("array")?;
-            this.stack.push(Element::Array);
-            Ok(())
-        })
+        self.handle_pending_collection()?;
+        self.pending_collection = Some(PendingCollection::Array);
+        Ok(())
     }
 
     fn write_start_dictionary(&mut self, _len: Option<u64>) -> Result<(), Error> {
-        self.write_value_event(EventKind::StartDictionary, |this| {
-            this.start_element("dict")?;
-            this.stack.push(Element::Dictionary);
-            Ok(())
-        })
+        self.handle_pending_collection()?;
+        self.pending_collection = Some(PendingCollection::Dictionary);
+        Ok(())
     }
 
     fn write_end_collection(&mut self) -> Result<(), Error> {
         self.write_event(|this| {
-            match (this.stack.pop(), this.expecting_key) {
+            match this.pending_collection.take() {
+                Some(PendingCollection::Array) => {
+                    this.xml_writer
+                        .write_event(XmlEvent::Empty(BytesStart::borrowed_name(b"array")))?;
+                    this.expecting_key = this.stack.last() == Some(&Element::Dictionary);
```

Comment:
I don't think this, or the copy below is needed. `pending_collection` will only be `Some` if this is called straight after `write_start_dictionary` or `write_start_array` in which case the value of `expecting_key` will have remained unchanged so the existing value is always valid.

Reply:
No need to change this now by the way.

Reply:
I'd need to look into what is happening here in more detail to understand why it is how it is, but with those statements removed the tests fail.